File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/RpcFlow.java
/*
 * Copyright (c) 2026 Super Durable, Inc.
 *
 * Licensed under the Super Durable Source License 1.0.
 * You may not use this file except in compliance with the License.
 * See the LICENSE file in the repository root.
 *
 * SPDX-License-Identifier: LicenseRef-Super-Durable-1.0
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.AttributeIndex;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDef;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.Wait;

import java.util.Arrays;
import java.util.List;

class RpcFlow implements Flow<Integer> {
    final Channel<Void> internal = Channel.define("rpc-internal", Void.class);
    final Attribute<String> data = Attribute.define("rpc-data", String.class);
    final Attribute<String> keyword = Attribute.define(
            "rpc-keyword",
            String.class,
            new AttributeIndex(AttributeIndex.Type.KEYWORD));
    private final RpcFirstStep first = new RpcFirstStep();
    private final RpcSecondStep second = new RpcSecondStep();

    @Override
    public List<StepDef> getSteps() {
        return Arrays.asList(
                StepDef.startStep(first),
                StepDef.nonStartStep(second));
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(data, keyword, internal);
    }

    @RPC
    public void noPersistence(final Context context) {
        internal.publish(context, null);
    }

    @RPC
    public RPCResult<Long> functionOne(final Context context, final String input) {
        data.set(context, input);
        keyword.set(context, input);
        return RPCResult.of(1L, StepMovement.of(second, 0));
    }

    @RPC
    public RPCResult<Long> functionZero(final Context context) {
        return RPCResult.of(1L, StepMovement.of(second, 0));
    }

    @RPC
    public void procedureOne(final Context context, final String input) {
        data.set(context, input);
    }

    @RPC
    public void procedureZero(final Context context) {
        internal.publish(context, null);
    }

    @RPC
    public RPCResult<Long> readOnly(final Context context, final String input) {
        return RPCResult.of((long) input.length());
    }

    @RPC
    public void setData(final Context context, final String input) {
        data.set(context, input);
    }

    @RPC
    public RPCResult<String> getData(final Context context) {
        return RPCResult.of(data.get(context));
    }

    @RPC
    public void setKeyword(final Context context, final String input) {
        keyword.set(context, input);
    }

    @RPC
    public RPCResult<String> getKeyword(final Context context) {
        return RPCResult.of(keyword.get(context));
    }

    final class RpcFirstStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer input) {
            return Wait.anyOf(internal.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            return StepDecision.goTo(second, 0);
        }
    }

    static final class RpcSecondStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            return StepDecision.gracefulComplete(input + 1);
        }
    }
}

File: sdk-typescript/test/integ/rpc_flow.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import {
  Attribute,
  Channel,
  IndexType,
  StepList,
  Wait,
  doubleCodec,
  goTo,
  gracefulComplete,
  optionalCodec,
  rpc,
  stringCodec,
  voidCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
  type Step,
  type StepDecision,
} from "../../src/index.js";

class RpcSecondStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public getStepType(): string {
    return "RpcSecondStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return gracefulComplete(2);
  }
}

class RpcFirstStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(
    private readonly internal: Channel<void>,
    private readonly second: RpcSecondStep,
  ) {}

  public getStepType(): string {
    return "RpcFirstStep";
  }

  public waitFor(_context: Context, _input: number): Wait {
    return Wait.anyOf(this.internal.forOne());
  }

  public execute(_context: Context, _input: number): StepDecision {
    return goTo(this.second, 0);
  }
}

export class RpcFlow implements Flow<number> {
  public static readonly RPC_OUTPUT = 100;
  public static readonly HARDCODED_VALUE = "random-string";
  public readonly internal = new Channel("rpc-internal", voidCodec);
  public readonly data = new Attribute("rpc-data", optionalCodec(stringCodec));
  public readonly keyword = new Attribute("CustomKeywordField", optionalCodec(stringCodec), {
    type: IndexType.KEYWORD,
  });
  public readonly integer = new Attribute("CustomIntField", doubleCodec, {
    type: IndexType.INT,
  });
  private readonly second = new RpcSecondStep();
  private readonly first = new RpcFirstStep(this.internal, this.second);

  public getFlowType(): string {
    return "RpcFlow";
  }

  public getSteps() {
    return StepList.startStep(this.first).otherSteps(this.second);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return {
      attributes: [this.data, this.keyword, this.integer],
      channels: [this.internal],
    };
  }

  @rpc()
  public noPersistence(context: Context): void {
    this.requireContext(context);
    this.internal.publish(context, undefined);
  }

  @rpc({ inputCodec: stringCodec, outputCodec: doubleCodec })
  public functionOne(context: Context, input: string): RPCResult<number> {
    this.requireContext(context);
    this.data.set(context, undefined);
    this.data.set(context, input);
    this.keyword.set(context, input);
    this.integer.set(context, RpcFlow.RPC_OUTPUT);
    this.internal.publish(context, undefined);
    return { output: RpcFlow.RPC_OUTPUT };
  }

  @rpc({ outputCodec: doubleCodec })
  public functionZero(_context: Context): RPCResult<number> {
    this.requireContext(_context);
    this.data.set(_context, RpcFlow.HARDCODED_VALUE);
    this.keyword.set(_context, RpcFlow.HARDCODED_VALUE);
    this.integer.set(_context, RpcFlow.RPC_OUTPUT);
    this.internal.publish(_context, undefined);
    return { output: RpcFlow.RPC_OUTPUT };
  }

  @rpc({ inputCodec: stringCodec })
  public procedureOne(context: Context, input: string): void {
    this.requireContext(context);
    this.data.set(context, input);
    this.keyword.set(context, input);
    this.integer.set(context, RpcFlow.RPC_OUTPUT);
    this.internal.publish(context, undefined);
  }

  @rpc()
  public procedureZero(context: Context): void {
    this.requireContext(context);
    this.data.set(context, RpcFlow.HARDCODED_VALUE);
    this.keyword.set(context, RpcFlow.HARDCODED_VALUE);
    this.integer.set(context, RpcFlow.RPC_OUTPUT);
    this.internal.publish(context, undefined);
  }

  @rpc({ inputCodec: stringCodec, outputCodec: doubleCodec })
  public readOnly(_context: Context, input: string): RPCResult<number> {
    this.requireContext(_context);
    return { output: RpcFlow.RPC_OUTPUT };
  }

  @rpc({ inputCodec: optionalCodec(stringCodec) })
  public setData(context: Context, input: string | undefined): void {
    this.requireContext(context);
    this.data.set(context, input);
  }

  @rpc({ outputCodec: optionalCodec(stringCodec) })
  public getData(context: Context): RPCResult<string | undefined> {
    this.requireContext(context);
    return { output: this.data.get(context) };
  }

  @rpc({ inputCodec: optionalCodec(stringCodec) })
  public setKeyword(context: Context, input: string | undefined): void {
    this.requireContext(context);
    this.keyword.set(context, input);
  }

  @rpc({ outputCodec: optionalCodec(stringCodec) })
  public getKeyword(context: Context): RPCResult<string | undefined> {
    this.requireContext(context);
    return { output: this.keyword.get(context) };
  }

  private requireContext(context: Context): void {
    if (context.flowId === "" || context.runId === "") {
      throw new Error("invalid RPC context");
    }
  }
}
